<a href="https://colab.research.google.com/github/chandan110791/HindiSpeechRecognition/blob/main/Whisper_LM_fusion_final.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Fusing LM with Whisper for lower WER
The aim is to fuse a BPE-level LM scores with the generated tokens scores while beam-search decoding in Whisper.

## **MILESTONE 1**:
Instantiate a Language Model to be integrated with Whisper.
The chosen LM is an n-gram language model, trained with [KenLM](https://github.com/kpu/kenlm) library.

### Step 1:
Write code to run an already available LM in a standalone manner and be able to give a score any input sequence.
**Chosen Model**: [Riva ASR Hindi LM](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/tao/models/speechtotext_hi_in_lm/files?version=deployable_v3.1)


Download and build the KenLM toolkit

In [ ]:
!wget -O - https://kheafield.com/code/kenlm.tar.gz | tar xz
!mkdir kenlm/build && cd kenlm/build && cmake .. && make -j2
!ls kenlm/build/bin

--2023-11-13 11:27:39--  https://kheafield.com/code/kenlm.tar.gz
Resolving kheafield.com (kheafield.com)... 35.196.63.85
Connecting to kheafield.com (kheafield.com)|35.196.63.85|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 491888 (480K) [application/x-gzip]
Saving to: ‘STDOUT’

-                   100%[===================>] 480.36K  1.67MB/s    in 0.3s    

2023-11-13 11:27:40 (1.67 MB/s) - written to stdout [491888/491888]

CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 3.5 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell
  CMake that the project does not need compatibility with older versions.


-- The C compiler identification is GNU 11.4.0
-- The CXX compiler identification is GNU 11.4.0
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /usr/bin/cc - skipped
-- Detectin

Download the KenLM Python library

❗❗**Don't forget to restart the runtime after running this cell** ❗❗


In [ ]:
!pip install https://github.com/kpu/kenlm/archive/master.zip

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 553.6/553.6 kB 6.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for kenlm: filename=kenlm-0.2.0-cp310-cp310-linux_x86_64.whl size=3184345 sha256=6cc9a5db8180947baa0ae0b8fd35c040b77ff8e307bf437b2fafd6295101aa79
  Stored in directory: /tmp/pip-ephem-wheel-cache-9az3r8uo/wheels/a5/73/ee/670fbd0cee8f6f0b21d10987cb042291e662e26e1a07026462
Successfully built kenlm


## ***Pointer 1:Please add comments on the model being downloaded and whats its purpose ***

Downloading the Hindi ASR n-gram language model from Nvidia which can be found [here](https://catalog.ngc.nvidia.com/orgs/nvidia/teams/tao/models/speechtotext_hi_in_lm/files?version=deployable_v3.1)

This will be used for fusion with Whisper.
I uploaded the binary version to a shareable location in my Gdrive.

In [ ]:
# Download the binary LM from Gdrive
import gdown
url = "https://drive.google.com/uc?id=1-AspJVZRXcrFMuLKx8C4N7uo9PnQytcB"
output = "language_model_3p0.bin"
gdown.download(url, output, quiet=False)

Downloading...
From: https://drive.google.com/uc?id=1-AspJVZRXcrFMuLKx8C4N7uo9PnQytcB
To: /content/language_model_3p0.bin
100%|██████████| 280M/280M [00:01<00:00, 147MB/s]


'language_model_3p0.bin'

In [ ]:
import kenlm
model = kenlm.LanguageModel('/content/language_model_3p0.bin')
print("This is a {}-gram model".format(model.order))

This is a 4-gram model


In [ ]:
# Below are 2 pairs of sentences that sound exactly the same in hindi but one of them is incorrect (lexically or semantically)
# Generated using Bing Chat
book_correct = "मुझे यह किताब पसंद है।"
book_incorrect = "मुझे यह किताब पसन्द है।"

correct_score = model.score(book_correct)
incorrect_score = model.score(book_incorrect)
assert correct_score > incorrect_score
print(correct_score, incorrect_score)

-20.935253143310547 -21.663846969604492


In [ ]:
sings_correct = "वह बहुत अच्छा गाता है।"
sings_incorrect = "वह बहुत अच्छा घाता है।"


correct_score = model.score(sings_correct)
incorrect_score = model.score(sings_incorrect)
assert correct_score > incorrect_score
print(correct_score, incorrect_score)

-19.827430725097656 -22.76061248779297


## ***Pointer 2:Please elaborate on the model being downloaded and whats its purpose . Also , why are we doing it two times  ***

This is unnecessary and can be skipped.

In [ ]:
# # download the original arpa LM file for inspection
# url = "https://drive.google.com/uc?id=1-4xQ3YCtsyONtpccGjOD1s9FtHqBX7RL"
# output = "language_model_3p0.arpa"
# gdown.download(url, output, quiet=False)

Downloading...
From: https://drive.google.com/uc?id=1-4xQ3YCtsyONtpccGjOD1s9FtHqBX7RL
To: /content/language_model_3p0.arpa
100%|██████████| 3.18G/3.18G [00:33<00:00, 95.6MB/s]


'language_model_3p0.arpa'

$$:Please elaborate if we need to prepare a new arpa file for our model

We can use this model for the time being, and finetune it later with the text data from our training set.
However, I think it's not likely to come up with a better model than this, as this model is built by a team in Nvidia and it's probable that our training set is included in training this model.

In [ ]:
# inspect the last 20 lines inside the LM source
!tail -20 language_model_3p0.arpa

-1.2548329	<s> दोनों ही झल्लाये
-0.2760634	<s> चौधरी के अशुभचिंतकों
-0.04973512	डालियों पर बैठी शुकमंडली
-0.07060567	मनुष्यों को उन्हें बेमुरौवत
-0.049646165	और कड़क कर बोलेमेरी
-0.04038189	निराश हो कर कहानहीं
-0.08863469	पड़ते ही वह अव्यवस्थितचित्त
-0.19321889	दोनों पक्षों से सवालजवाब
-0.051110353	झगड़ू साहु ने कहासमझू
-0.20675866	करें तो उनकी भलमनसी
-0.04876329	नीति को सराहता थाइसे
-0.06436408	<s> मित्रता की मुरझायी
-0.23735626	की गहराई से उपजतें
-0.17502813	पूर्णता की ओर बढातें
-0.18197767	जहाँ से अच्छा हिन्दोसिताँ
-0.04437429	हैं इसकी यह गुलसिताँ
-0.06926097	संतरी हमारा वह पासबाँ
-0.09434804	जिनके दम से रश्कएजनाँ

\end\


In [ ]:
# some useful KenLM commands for future reference
# generate binary
# !kenlm/build/bin/build_binary dataset_tokenized_3gram.arpa dataset_tokenized_3gram.binary
# create a new LM
# !kenlm/build/bin/lmplz -o 3 --text dataset_tokenized.txt --arpa dataset_tokenized_3gram.arpa --discount_fallback
# !tail -20 dataset_tokenized_3gram.arpa

# Integrating the LM with Whisper

In [ ]:
# !pip install openai-whisper
!pip install https://github.com/chandan110791/hindiWhisper/archive/master.zip

     \ 7.4 MB 16.9 MB/s 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.3/63.3 MB 8.5 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 74.0 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 153.0/153.0 kB 17.2 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
INFO: pip is looking at multiple versions of torch to determine which version is compatible with other requirements. This could take a while.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 670.2/670.2 MB 963.9 kB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.7/23.7 MB 55.7 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.6/823.6 kB 52.9 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━

In [ ]:
import whisper
import torch
import kenlm

In [ ]:
model = whisper.load_model("small")

100%|████████████████████████████████████████| 461M/461M [00:04<00:00, 112MiB/s]


In [ ]:
# Download a sample audio file from Gdrive
import gdown
url = "https://drive.google.com/uc?id=1kKeSvrZo8z5Rsp1q-h3GXpG7vHctKMcG"
output = "sample.wav"
gdown.download(url, output, quiet=False)

transcription = "ब्रूड बॉक्स लैंगस्ट्रॉथ छत्ते का एक अनिवार्य हिस्सा है।"

Downloading...
From: https://drive.google.com/uc?id=1kKeSvrZo8z5Rsp1q-h3GXpG7vHctKMcG
To: /content/sample.wav
100%|██████████| 197k/197k [00:00<00:00, 77.9MB/s]


In [ ]:
audio = whisper.load_audio("/content/sample.wav")
audio = whisper.pad_or_trim(audio)
mel = whisper.log_mel_spectrogram(audio).to(model.device)


In [ ]:
# Download a sample audio file from Gdrive
import gdown
url = "https://drive.google.com/uc?id=1kKeSvrZo8z5Rsp1q-h3GXpG7vHctKMcG"
output = "sample.wav"
gdown.download(url, output, quiet=False)

transcription = "ब्रूड बॉक्स लैंगस्ट्रॉथ छत्ते का एक अनिवार्य हिस्सा है।"


In [ ]:
audio = whisper.load_audio("/content/sample.wav")
audio = whisper.pad_or_trim(audio)
mel = whisper.log_mel_spectrogram(audio).to(model.device)
options = whisper.DecodingOptions(fp16 = False, beam_size=5, without_timestamps=True, language="hi")
result = whisper.decode(model, mel, options)
baseline = result.text
baseline

## Baseline Whisper
Without nbest or LM integration

In [ ]:

options = whisper.DecodingOptions(fp16 = False, beam_size=5, without_timestamps=True, language="hi")
result = whisper.decode(model, mel, options)
baseline = result.text
baseline

'ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा है'

# Decoding with LM integration

In [ ]:
# adding the LM
import gdown
url = "https://drive.google.com/uc?id=1-AspJVZRXcrFMuLKx8C4N7uo9PnQytcB"
output = "language_model_3p0.bin"
gdown.download(url, output, quiet=False)

Downloading...
From: https://drive.google.com/uc?id=1-AspJVZRXcrFMuLKx8C4N7uo9PnQytcB
To: /content/language_model_3p0.bin
100%|██████████| 280M/280M [00:04<00:00, 67.3MB/s]


'language_model_3p0.bin'

$$:

## ***Pointer 4: Getting an error here while trying to run the code . Elaborate a bit on the what happening here . Can you please refer to the function with changes done  ***

The error was due to the fact that the imported whisper here is the original one, which doesn't include the changed that we did to the decoding options.

Replaced this:
```
!pip install openai-whisper
```
with this:
```
!pip install https://github.com/chandan110791/hindiWhisper/archive/master.zip
```

and the error is fixed.

In [ ]:
options = whisper.DecodingOptions(fp16 = False, withlm=True, beam_size=5,
        patience=1.0, lm_path="/content/language_model_3p0.bin", lm_alpha=1.0, lm_beta=0.0,
        without_timestamps=True, language="hi")
decoding_withLM = whisper.decode(model, mel, options)


In [ ]:
decoding_withLM.text

'ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा है'

# Decoding with nbest (beam search)

In [ ]:
options = whisper.DecodingOptions(fp16 = False, beam_size=5, return_nbest = True, without_timestamps=True, language="hi")
nbest = whisper.decode(model, mel, options)


for candidate in nbest:
  print(candidate.text, candidate.avg_logprob)

ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा है -0.45291578358617324
ब्रूद बाँक्ष लांश्टोट छत्ते का एक अनिवार्य हिस्सा है -0.45523316981428763
ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा है. -0.47296941078315347
ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा हैं -0.48318856449450476
ब्रूद बाँक्ष लांश्टोट छत्टे का एक अनिवार्य हिस्सा है -0.46669308344523114


### Adding LM rescoring

In [ ]:
lm_model = kenlm.LanguageModel('/content/language_model_3p0.bin')
print("This is a {}-gram model".format(lm_model.order))

This is a 4-gram model


In [ ]:
nbest_with_lm_score = [(c.text, c.avg_logprob, lm_model.score(c.text)) for c in nbest]
nbest_with_lm_score

[('ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा है',
  -0.45291578358617324,
  -36.06303787231445),
 ('ब्रूद बाँक्ष लांश्टोट छत्ते का एक अनिवार्य हिस्सा है',
  -0.45523316981428763,
  -36.06303787231445),
 ('ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा है.',
  -0.47296941078315347,
  -44.10427474975586),
 ('ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा हैं',
  -0.48318856449450476,
  -37.30411148071289),
 ('ब्रूद बाँक्ष लांश्टोट छत्टे का एक अनिवार्य हिस्सा है',
  -0.46669308344523114,
  -37.54087829589844)]

In [ ]:
lm_weight = 0.01
combined_scores = [(text, whisper_score + lm_score*lm_weight) for text, whisper_score, lm_score in nbest_with_lm_score]
combined_scores.sort(key=lambda t: t[1], reverse=True)
combined_scores

[('ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा है', -0.8135461623093178),
 ('ब्रूद बाँक्ष लांश्टोट छत्ते का एक अनिवार्य हिस्सा है', -0.8158635485374321),
 ('ब्रूद बाँक्ष लांश्टोट छत्टे का एक अनिवार्य हिस्सा है', -0.8421018664042155),
 ('ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा हैं', -0.8562296793016337),
 ('ब्रूद बाँक्ष लंश्टोट छत्ते का एक अनिवार्य हिस्सा है.', -0.9140121582807121)]

# Decoding with nbest (best of N hypothesis)

In [ ]:
options = whisper.DecodingOptions(fp16 = False, best_of=10, return_nbest=True, without_timestamps=True, temperature=0.3, language="hi")
nbest_best_of_n_hyp = whisper.decode(model, mel, options)


In [ ]:
for candidate in nbest_best_of_n_hyp:
  print(candidate.text, candidate.avg_logprob)

ब्रूद बाँच लंश्टोट छत्ते का एक अनिवार्य हिस्सा है -0.45742596898760113
ब्रूद बाओ्छ लंश्टोट छत्ते का एक अनिवार्ये हिस्चा है -0.5202431113032971
ब्रूड़ ब्रूड़ लंच्टोट चद्ते का एक अनीवार्य लिए हिस्चा है -0.5065439448637121
ब्रुध बाँच लंश्टोट चते का एक अनिवार्य हिस morally part of the Landstorch Chhattey. -0.9538334877260269
ब्रुद बाँच लांश्टोट छद्टे का एक अनिवार्य हिस्चा है -0.5101523081461589
ब्रूद बाँच लंच्टोड छत्टे का एक अनिवार्य लिए हिस्सा है -0.4865361798194147
ब्रूद बाँच लांश्टोट छत्ते का एक अनीवार्य हिस्था है -0.49694071144893254
ब्रूद बाँच लन्च्टोट चते का एक अनिवार्य हिस्चा है -0.5291237149919782
ब्रूद भोग्ष लंग स्थ्टोड शथ्टे का एक अनिवार्य हिस्सा है -0.5426437135726686
ब्रूद बाँच लंश्टोट छद्टे का एक अनिवारे हिस्चा है -0.49345003325363684


In [ ]:
lm_model = kenlm.LanguageModel('/content/language_model_3p0.bin')
nbest_with_lm_score = [(c.text, c.avg_logprob, lm_model.score(c.text)) for c in nbest_best_of_n_hyp]
nbest_with_lm_score


[('ब्रूद बाँच लंश्टोट छत्ते का एक अनिवार्य हिस्सा है',
  -0.45742596898760113,
  -35.82866668701172),
 ('ब्रूद बाओ्छ लंश्टोट छत्ते का एक अनिवार्ये हिस्चा है',
  -0.5202431113032971,
  -49.43550491333008),
 ('ब्रूड़ ब्रूड़ लंच्टोट चद्ते का एक अनीवार्य लिए हिस्चा है',
  -0.5065439448637121,
  -54.91597366333008),
 ('ब्रुध बाँच लंश्टोट चते का एक अनिवार्य हिस morally part of the Landstorch Chhattey.',
  -0.9538334877260269,
  -81.96270751953125),
 ('ब्रुद बाँच लांश्टोट छद्टे का एक अनिवार्य हिस्चा है',
  -0.5101523081461589,
  -46.626407623291016),
 ('ब्रूद बाँच लंच्टोड छत्टे का एक अनिवार्य लिए हिस्सा है',
  -0.4865361798194147,
  -46.603515625),
 ('ब्रूद बाँच लांश्टोट छत्ते का एक अनीवार्य हिस्था है',
  -0.49694071144893254,
  -49.201133728027344),
 ('ब्रूद बाँच लन्च्टोट चते का एक अनिवार्य हिस्चा है',
  -0.5291237149919782,
  -46.626407623291016),
 ('ब्रूद भोग्ष लंग स्थ्टोड शथ्टे का एक अनिवार्य हिस्सा है',
  -0.5426437135726686,
  -42.747047424316406),
 ('ब्रूद बाँच लंश्टोट छद्टे का एक अनिव

In [ ]:
lm_weight = 0.01
combined_scores_bestofNSampling = [(text, whisper_score + lm_score*lm_weight) for text, whisper_score, lm_score in nbest_with_lm_score]
combined_scores_bestofNSampling.sort(key=lambda t: t[1], reverse=True)
combined_scores_bestofNSampling

[('ब्रूद बाँच लंश्टोट छत्ते का एक अनिवार्य हिस्सा है', -0.8157126358577182),
 ('ब्रूद बाँच लंच्टोड छत्टे का एक अनिवार्य लिए हिस्सा है',
  -0.9525713360694148),
 ('ब्रूद भोग्ष लंग स्थ्टोड शथ्टे का एक अनिवार्य हिस्सा है',
  -0.9701141878158327),
 ('ब्रुद बाँच लांश्टोट छद्टे का एक अनिवार्य हिस्चा है', -0.976416384379069),
 ('ब्रूद बाँच लांश्टोट छत्ते का एक अनीवार्य हिस्था है', -0.9889520487292061),
 ('ब्रूद बाँच लन्च्टोट चते का एक अनिवार्य हिस्चा है', -0.9953877912248883),
 ('ब्रूद बाँच लंश्टोट छद्टे का एक अनिवारे हिस्चा है', -1.0002398129167227),
 ('ब्रूद बाओ्छ लंश्टोट छत्ते का एक अनिवार्ये हिस्चा है', -1.014598160436598),
 ('ब्रूड़ ब्रूड़ लंच्टोट चद्ते का एक अनीवार्य लिए हिस्चा है',
  -1.0557036814970129),
 ('ब्रुध बाँच लंश्टोट चते का एक अनिवार्य हिस morally part of the Landstorch Chhattey.',
  -1.7734605629213394)]

# **RUN THE NOTEBOOK STARTING HERE**

# The next sections include:
- Importing a finetuned huggingface model to our codebase
- Running the evaluation of the 4 decoding variants on the imported dataset
- Computing the WER and CER for each of the decoding variants

## The decoding strategies are:
1. baseline decoding without LM
2. Deep fusion of the LM with the token probabilities during beam search decoding (try to find the optimal value for lm_alpha)
3. Shallow fusion by rescoring the N best candidates generated through beam search (try to find the optimal value for lm_weight)
4. Shallow fusion by rescoring the N best candidates generated through greedy decoding using best of N sampling (try to find the optimal value for lm_weight and temperature)





In [1]:
!pip install https://github.com/chandan110791/hindiWhisper/archive/master.zip
!pip install https://github.com/kpu/kenlm/archive/master.zip
!pip install transformers
!pip install datasets>=2.6.1
!pip install git+https://github.com/huggingface/transformers
!pip install librosa
# !pip install evaluate>=0.30
!pip install jiwer
!pip install gradio
!pip install accelerate -U

     / 7.6 MB 12.5 MB/s 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 63.3/63.3 MB 29.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.0/2.0 MB 94.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 153.0/153.0 kB 22.9 MB/s eta 0:00:00
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
INFO: pip is looking at multiple versions of torch to determine which version is compatible with other requirements. This could take a while.
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 670.2/670.2 MB 2.3 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 23.7/23.7 MB 59.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 823.6/823.6 kB 73.2 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━

In [2]:
# Download the binary LM from Gdrive
import gdown
url = "https://drive.google.com/uc?id=1-AspJVZRXcrFMuLKx8C4N7uo9PnQytcB"
output = "language_model_3p0.bin"
gdown.download(url, output, quiet=False)

Downloading...
From: https://drive.google.com/uc?id=1-AspJVZRXcrFMuLKx8C4N7uo9PnQytcB
To: /content/language_model_3p0.bin
100%|██████████| 280M/280M [00:02<00:00, 93.5MB/s]


'language_model_3p0.bin'

$$

$$:

## ***Pointer 5: its looking for a file named as pytorchmodel.bin , it does not exists hence note  able to integrate the model : CKSINGH/whisper-small-hi-iiib tuned at : https://colab.research.google.com/github/chandan110791/HindiSpeechRecognition/blob/main/fine_tune_whisper_iiitb.ipynb#scrollTo=d7030622-caf7-4039-939b-6195cdaa2585  . Any inputs ***

The reason for this is that the model file after finetuning must be pickled:
so pytorch_model.bin should be found in this path:
https://huggingface.co/CKSINGH/whisper-small-hi-iiib/tree/main

Will use this one https://huggingface.co/sanchit-gandhi/whisper-small-hi/tree/main to make the required changes to the code.

## In the next cell, you need to change the repo id to your repo after generating the pytorch_model.bin (pickle format of the finetuned model)

In [3]:
import librosa
import soundfile as sf

def resample_audio(input_file_path, output_file_path, orig_sr, target_sr):
    """
    Resample an audio file from original sample rate to target sample rate.

    Args:
    - input_file_path (str): Path to the input audio file.
    - output_file_path (str): Path where the resampled audio file will be saved.
    - orig_sr (int): Original sample rate of the audio file.
    - target_sr (int): Target sample rate to resample the audio file.
    """
    # Load the audio file
    audio, sr = librosa.load(input_file_path, sr=orig_sr)

    # Resample to target sample rate
    audio_resampled = librosa.resample(audio, orig_sr=sr, target_sr=target_sr)

    # Save the resampled audio
    sf.write(output_file_path, audio_resampled, target_sr)


In [4]:
import whisper
import kenlm
from transformers import WhisperForConditionalGeneration
import torch
import jiwer
from tqdm import tqdm
import os
import pandas as pd
# using pickle to serialize the map_dict
import pickle

from huggingface_hub import hf_hub_download
filename = "pytorch_model.bin"

#hf_hub_download(repo_id="CKSINGH/whisper-small-hi-iiib", filename=filename, local_dir="/content/")
#hf_hub_download(repo_id="CKSINGH/whisper-small-hi-firefox", filename=filename, local_dir="/content/")
#hf_hub_download(repo_id="sanchit-gandhi/whisper-small-hi", filename=filename, local_dir="/content/")
hf_hub_download(repo_id="CKSINGH/whisper-small-hi-graminVoice", filename=filename, local_dir="/content/")

pytorch_model.bin:   0%|          | 0.00/967M [00:00<?, ?B/s]

'/content/pytorch_model.bin'

# Here you need to change the model size from "small" to "medium" if you finetuned the whisper-medium (as the one in this repo: https://huggingface.co/CKSINGH/whisper-small-hi-iiib/tree/main)

In [5]:
# to enable verbose printing of exceptions (+ layers matching name)
DEBUG = True

# set to True if your custom model has been trained using DDP (multi-gpu)
# as in my case, in the custom HF model, keys have a prefix (model.)
# it should come from the fact that I have trained on a milti-gpu machine, using DDP
DDP_TRAINED = True

# if DDP we have to add a prefix to match with the HF state_dict
if DDP_TRAINED:
    PREFIX = "model."
else:
    PREFIX = ""

MODEL_SIZE = "small"

# the device where you're running this code
DEVICE = torch.device('cuda' if torch.cuda.is_available() else 'cpu')

# the name of the file with your fine-tuned model
FINETUNED_MODEL = "pytorch_model.bin"

# the name of the file for the serialized map_dict
# a different name, to avoid overwrite it
FILE_DICT = MODEL_SIZE + "_map_dict01.pkl"


In [6]:
def import_hf_model(finetuned_model, model_size, device, file_dict, debug=True):

  def has_numbers(inputString):
      return any(char.isdigit() for char in inputString)

  # next functions are used to make sanity checks for the mappings

  # get if it is encoder or decoder
  def extract_function(key_name):
      # encoder or decoder is the first part of the key
      first_part = key_name.split(".")[0]

      key_func = None
      if first_part in ["enconder", "decoder"]:
          key_func = first_part

      return key_func

  def extract_layer_num(key_name):
      # layer num is the third piece
      layer_num = None

      if has_numbers(key_name):
          layer_num = key_name.split(".")[2]

      return layer_num

  # check that the two keys are for layers
  # with the same function
  # (both encoder or both decoder)
  # and have the same layer number
  # this way we are super-safe (I think)
  def sanity_check(key1, key2):
      is_ok = True

      # check same func (encoder or decoder)
      func1 = extract_function(key1)
      func2 = extract_function(key2)

      if func1 != func2:
          print(f"Warning: layers seem to have different functions: {key1},{key2}")
          is_ok = False

      # check same layer_num
      layer1 = extract_layer_num(key1)
      layer2 = extract_layer_num(key2)

      if layer1 != layer2:
          print(f"Warning: layers seem to have different numbers: {key1},{key2}")
          is_ok = False

      return is_ok

  if not os.path.isfile(file_dict):
    # Vanilla means: not custom trained
    print()
    print("Loading vanilla Whisper model")
    model = whisper.load_model(model_size, device=device)

    print("Loading vanilla HF Model")
    hugging_face_model = WhisperForConditionalGeneration.from_pretrained(
        "openai/whisper-" + model_size
    ).to(device)

    # extract state-dict from both
    state_d_openai = model.state_dict()
    state_d_huggingface = hugging_face_model.model.state_dict()

    # build the mapping between keys...
    map_dict = {}
    print("Matching layers...")

    # for every layer in OpenAI model
    n_sanity_ok = 0

    #
    # here we're considering the cartesian product of the two state dict and try to match
    # rules applied:
    # 1. the two layers have the same shape
    # 2. the two layer have the same parameters' values
    # 3. we apply sanity check (see function above)
    #
    for k in tqdm(state_d_openai):
        # find a layer in the HF model, check with j
        for j in state_d_huggingface:
            # where parameters have same shape and same values
            if state_d_huggingface[j].shape == state_d_openai[k].shape:
                if torch.all(torch.eq(state_d_huggingface[j], state_d_openai[k])).item():
                    # found, register the mapping
                    map_dict[k] = j
                    # make some check and eventually print a warning
                    if sanity_check(k, j) == True:
                        n_sanity_ok += 1

                        # if you enable thsi print you can see the name of the layer
                        # chosen in the match and you will se that they have the same functions
                        if debug:
                            print(k, j)

                    break


    # check if we have matched every entry
    print("Check if we have matched every entry in state_dict...")
    print()
    print(f"Number of keys: {len(map_dict.keys())}")
    assert len(map_dict.keys()) == len(state_d_openai.keys()), "The match is not complete !"

    print(f"Number of sanity_check ok: {n_sanity_ok}")
    print()

    print("Match is complete !!!")
    print()


    # serialize the map_dict to file
    print("Serializing map_dict...")

    with open(file_dict, "wb") as f:
        pickle.dump(map_dict, f)
        f.close()

    print(f"map_dict saved as: {file_dict}...")
    print()

  else:
    # loading with match keys
    # restart from pickle file
    print("Reloading map_dict...")
    print()
    with open(file_dict, "rb") as f:
        map_dict = pickle.load(f)

  # loading fine-tuned dict
  print("Loading fine tuned dict...")

  # added map_location to handle the fact that the custom model has been trained on GPU
  state_dict_finetuned = torch.load(finetuned_model, map_location=torch.device(device))

  print(state_dict_finetuned.keys())
  # build the state_dict to be used
  # take the key name from standard (OpenAI) and the value from finetuned (HF)
  print("Rebuild the state dict...")
  new_state_dict = {}
  n_except = 0
  for k in tqdm(map_dict.keys()):
      try:
        # You must add "model." if you have used DDP in custom training
        # see DDP_TRAINED above
        # PREFIX is added to a HF fine-tuned 8with DDP). It is not in vanulla HF models
        new_state_dict[k] = state_dict_finetuned[PREFIX + map_dict[k]]
      except Exception as ex:
        n_except += 1

        if debug:
            print("exception")
            print(PREFIX + map_dict[k])




  msg_err = f"Rebuild state dict failed, {n_except} pick failed"
  assert n_except == 0, msg_err



  print()
  print("Loading the final model...")
  model.load_state_dict(new_state_dict)
  return model

In [38]:
model = ""

In [7]:
model = import_hf_model(finetuned_model=FINETUNED_MODEL, debug= False, model_size=MODEL_SIZE, device=DEVICE, file_dict=FILE_DICT)


Loading vanilla Whisper model


100%|███████████████████████████████████████| 461M/461M [00:07<00:00, 61.7MiB/s]


Loading vanilla HF Model


config.json:   0%|          | 0.00/1.97k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/967M [00:00<?, ?B/s]

generation_config.json:   0%|          | 0.00/3.84k [00:00<?, ?B/s]

Matching layers...


100%|██████████| 479/479 [00:01<00:00, 264.47it/s]


Check if we have matched every entry in state_dict...

Number of keys: 479
Number of sanity_check ok: 479

Match is complete !!!

Serializing map_dict...
map_dict saved as: small_map_dict01.pkl...

Loading fine tuned dict...
odict_keys(['model.encoder.conv1.weight', 'model.encoder.conv1.bias', 'model.encoder.conv2.weight', 'model.encoder.conv2.bias', 'model.encoder.embed_positions.weight', 'model.encoder.layers.0.self_attn.k_proj.weight', 'model.encoder.layers.0.self_attn.v_proj.weight', 'model.encoder.layers.0.self_attn.v_proj.bias', 'model.encoder.layers.0.self_attn.q_proj.weight', 'model.encoder.layers.0.self_attn.q_proj.bias', 'model.encoder.layers.0.self_attn.out_proj.weight', 'model.encoder.layers.0.self_attn.out_proj.bias', 'model.encoder.layers.0.self_attn_layer_norm.weight', 'model.encoder.layers.0.self_attn_layer_norm.bias', 'model.encoder.layers.0.fc1.weight', 'model.encoder.layers.0.fc1.bias', 'model.encoder.layers.0.fc2.weight', 'model.encoder.layers.0.fc2.bias', 'model.en

100%|██████████| 479/479 [00:00<00:00, 1219096.85it/s]


Loading the final model...


In [8]:
model

Whisper(
  (encoder): AudioEncoder(
    (conv1): Conv1d(80, 768, kernel_size=(3,), stride=(1,), padding=(1,))
    (conv2): Conv1d(768, 768, kernel_size=(3,), stride=(2,), padding=(1,))
    (blocks): ModuleList(
      (0-11): 12 x ResidualAttentionBlock(
        (attn): MultiHeadAttention(
          (query): Linear(in_features=768, out_features=768, bias=True)
          (key): Linear(in_features=768, out_features=768, bias=False)
          (value): Linear(in_features=768, out_features=768, bias=True)
          (out): Linear(in_features=768, out_features=768, bias=True)
        )
        (attn_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
        (mlp): Sequential(
          (0): Linear(in_features=768, out_features=3072, bias=True)
          (1): GELU(approximate='none')
          (2): Linear(in_features=3072, out_features=768, bias=True)
        )
        (mlp_ln): LayerNorm((768,), eps=1e-05, elementwise_affine=True)
      )
    )
    (ln_post): LayerNorm((768,), eps=1e-0

In [ ]:
from huggingface_hub import notebook_login
# #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
notebook_login()

$$:

$$

$$:

## ***Pointer 6: We need to evaluate and calculate the WER,CER  on the Test data shared using our final model . Below are some boiler plates code to do the same  ***



In [8]:
#Prepare Test Data for calculating Wer and CER

import datasets
# import the load_dataset function
from datasets import load_dataset

# specify the URL directory and the data files
# load the dataset from the URL directory


datasets.config.DEFAULT_MAX_BATCH_SIZE = 10
#integrated_LM_dataset_1 = load_dataset("/content/datadownload.py")
integrated_LM_dataset_2 = load_dataset("/content/datadownload_iith.py")
#integrated_LM_dataset_3 = load_dataset("/content/graminVoiceDatadownload.py")

Generating txt.done.data.utf8 split: 0 examples [00:00, ? examples/s]

In [11]:
integrated_LM_dataset_2

DatasetDict({
    txt.done.data.utf8: Dataset({
        features: ['path', 'audio', 'sentence'],
        num_rows: 1000
    })
})

In [9]:
#integrated_LM_dataset_1 = integrated_LM_dataset_1.remove_columns(["age", "client_id", "down_votes", "gender", "locale", "path", "segment", "up_votes","accents","variant"])
integrated_LM_dataset_2 = integrated_LM_dataset_2.remove_columns(["path"])
#integrated_LM_dataset_3 = integrated_LM_dataset_3.remove_columns(["path"])


In [ ]:
integrated_LM_dataset_2

DatasetDict({
    txt.done.data.utf8: Dataset({
        features: ['audio', 'sentence'],
        num_rows: 1000
    })
})

In [10]:
#combined_dataset = datasets.concatenate_datasets([integrated_LM_dataset_1["test"]])
combined_dataset = datasets.concatenate_datasets([integrated_LM_dataset_2["txt.done.data.utf8"]])
#combined_dataset = datasets.concatenate_datasets([integrated_LM_dataset_3["text"]])


In [16]:
from datasets import load_dataset, concatenate_datasets, DatasetDict, load_metric

# Define the split ratios
#train_split = 0.75  # 80% of the data
validation_split = 0.50  # 10% of the data
test_split = 0.1  # 10% of the data

# Compute the number of samples for each split
num_samples = len(combined_dataset)
num_validation = int(validation_split * num_samples)
num_test = num_samples - num_validation  # Remaining 10%

# Split the combined dataset
validation_dataset = combined_dataset.select(indices=list(range(num_validation)))
test_dataset = combined_dataset.select(indices=list(range(num_validation, num_samples)))

# If you want to organize the split datasets in a DatasetDict for convenience:
split_test_datasets = DatasetDict({
    'validation': validation_dataset,
    'test': test_dataset
})

# Verify the resulting datasets
print(f'Validation Dataset: {len(validation_dataset)} samples')
print(f'Test Dataset: {len(test_dataset)} samples')

Validation Dataset: 500 samples
Test Dataset: 500 samples


In [ ]:
## you can use only 10% of the data to perform evaluation

Evaluation Metrics on the Test data

#correct the sampling rate

In [19]:

class customDataset(torch.utils.data.Dataset):
    """
    A simple class to wrap LibriSpeech and trim/pad the audio to 30 seconds.
    It will drop the last few seconds of a very small portion of the utterances.
    """
    original_sample_rate = 48000  # 48kHz
    target_sample_rate = 16000    # 16kHz

    def __init__(self, dataset, device=DEVICE):
        self.dataset = dataset
        self.device = device

    def __len__(self):
        return len(self.dataset)

    def __getitem__(self, item):
        # audio, sample_rate, text, _, _, _ = self.dataset[item]
        # assert sample_rate == 16000
        # audio = whisper.pad_or_trim(audio.flatten()).to(self.device)
        # mel = whisper.log_mel_spectrogram(audio)
        audio = self.dataset[item]['audio']
        sentence = self.dataset[item]['sentence']
        path = audio['path']
        original_sample_rate = 48000  # 48kHz
        target_sample_rate = 16000    # 16kHz
        resample_audio(path, path, original_sample_rate, target_sample_rate)
        array = audio['array']
        sampling_rate = audio['sampling_rate']
        audio = whisper.load_audio(path)
        audio = whisper.pad_or_trim(audio)
        mel = whisper.log_mel_spectrogram(audio).to(model.device)
        return (mel, sentence)

dataset = customDataset(split_test_datasets["test"])
loader = torch.utils.data.DataLoader(dataset, batch_size=1)

In [ ]:
dataset[1]

(tensor([[-0.1111, -0.3936, -0.3936,  ..., -0.3936, -0.3936, -0.3936],
         [-0.1780, -0.2980, -0.3936,  ..., -0.3936, -0.3936, -0.3936],
         [ 0.0772, -0.2939, -0.2480,  ..., -0.3936, -0.3936, -0.3936],
         ...,
         [-0.3936, -0.3936, -0.3936,  ..., -0.3936, -0.3936, -0.3936],
         [-0.3936, -0.3936, -0.3936,  ..., -0.3936, -0.3936, -0.3936],
         [-0.3936, -0.3936, -0.3936,  ..., -0.3936, -0.3936, -0.3936]],
        device='cuda:0'),
 'पालन करने को लेकर आम लोगों में जागरूकता नहीं यहाँ के लोग यातायात नियमों के पालन करने को लेकर गंभीर नही है इसी वजह से आजकल सड़क हादसों में')

In [20]:
def decode_baseline(model, mel, beam_size):
  """
    This function performs the transcription with Whisper to provide a baseline without LM fusion

    - model: The Whisper model used for transcription.
    - mel: Represents Mel-spectrogram data, likely input features for the model.
    - beam_size: An integer specifying the size of the beam for beam search decoding.
  """
  if mel.shape[0] != 1:
    mel = torch.unsqueeze(mel, 0)
  options = whisper.DecodingOptions(fp16 = False, beam_size=beam_size, without_timestamps=True, language="hi")
  result = whisper.decode(model, mel, options)
  result = [r.text for r in result]
  return result

def decode_deep_fusion(model, mel, beam_size, lm_path, lm_alpha):
  """
    This function performs the deep fusion of the LM with Whisper during the beam search decoding step

    - model: The Whisper model used for transcription.
    - mel: Represents Mel-spectrogram data, likely input features for the model.
    - beam_size: An integer specifying the size of the beam for beam search decoding.
    - lm_path: A string representing the path to the language model file used for fusion.
    - lm_alpha: A numerical value representing the weight assigned to the language model scores during deep fusion.
  """
  if mel.shape[0] != 1:
    mel = torch.unsqueeze(mel, 0)
  options = whisper.DecodingOptions(fp16 = False, withlm=True, beam_size=beam_size,
        patience=1.0, lm_path=lm_path, lm_alpha=lm_alpha, lm_beta=0.0,
        without_timestamps=True, language="hi")
  result = whisper.decode(model, mel, options)
  result = [r.text for r in result]
  return result

def decode_shallow_fusion_beam_search(model, mel, beam_size, lm_path, lm_weight, debug=False):
  """
    this function performs shallow fusion using best of N hypothesis (decoding)
    by combining the scores of whisper and the language model score (which gets weighted by the lm_weight factor)

    - model: The Whisper model used for transcription.
    - mel: Represents Mel-spectrogram data, likely input features for the model.
    - beam_size: An integer specifying the size of the beam for beam search decoding.
    - lm_path: A string representing the path to the language model file used for fusion.
    - lm_weight: A numerical value representing the weight assigned to the language model scores during shallow fusion.
    - debug: A boolean flag (optional) indicating whether to print debug information (default is False).
      Useful for inspecting the outputs with different lm_weights for finding the optimal value for lm_weight

  """
  # if testing with a single utterance without a dataloader
  if mel.shape[0] != 1:
    mel = torch.unsqueeze(mel, 0)
  options = whisper.DecodingOptions(fp16 = False, beam_size=beam_size, return_nbest = True, without_timestamps=True, language="hi")
  nbests = whisper.decode(model, mel, options)

  lm_model = kenlm.LanguageModel(lm_path)
  combined_scores = []

  for nbest in nbests:
    nbest_with_lm_score = [(c.text, c.avg_logprob, lm_model.score(c.text)) for c in nbest]
    combined_score = [(text, (whisper_score) + (lm_score*lm_weight), (whisper_score), (lm_score)) for text, whisper_score, lm_score in nbest_with_lm_score]
    combined_score.sort(key=lambda t: t[1], reverse=True)
    combined_scores.append(combined_score)


  if debug:
    print(combined_scores)
  # text, final_score, whisper_score, lm_score = combined_scores[0]
  # return the highest score element for each input in the batch
  result = []
  for datapoint in combined_scores:
    combined_score = datapoint[0]
    hyp, comb_score, w_score, lm_score = combined_score
    result.append(hyp)
  # result = [hyp for hyp, comb_score, w_score, lm_score in combined_scores]
  #return text
  return result

# def decode_shallow_fusion_nbest(model, mel, best_of, lm_path, temperature, lm_weight, debug=False):
#   """
#     this function performs shallow fusion using best of N hypothesis (greedy decoding)
#     by combining the scores of whisper and the language model score (which gets weighted by the lm_weight factor)

#     - model: The Whisper model used for transcription.
#     - mel: Represents Mel-spectrogram data, likely input features for the model.
#     - best_of: An integer specifying the number of best hypotheses to consider during decoding.
#     - lm_path: A string representing the path to the language model used for fusion.
#     - temperature: A numerical value indicating the temperature parameter used during sampling. Higher temperature corresponds to more variation in the n best list
#     - lm_weight: A numerical value representing the weight assigned to the language model scores during shallow fusion.
#     - debug: A boolean flag (optional) indicating whether to print debug information (default is False).
#       Useful when finding the optimal value for the lm_weight
#   """
#   # if testing with a single utterance without a dataloader
#   if mel.shape[0] != 1:
#     mel = torch.unsqueeze(mel, 0)
#   options = whisper.DecodingOptions(fp16 = False, best_of=best_of, return_nbest=True, without_timestamps=True, temperature=temperature, language="hi")
#   nbests = whisper.decode(model, mel, options)

#   lm_model = kenlm.LanguageModel(lm_path)
#   combined_scores = []

#   for nbest in nbests:
#     nbest_with_lm_score = [(c.text, c.avg_logprob, lm_model.score(c.text)) for c in nbest]
#     combined_score = [(text, (whisper_score) + (lm_score*lm_weight), (whisper_score), (lm_score)) for text, whisper_score, lm_score in nbest_with_lm_score]
#     combined_score.sort(key=lambda t: t[1], reverse=True)
#     combined_scores.append(combined_score)
#   if debug:
#     print(combined_scores)
#   # text, final_score, whisper_score, lm_score = combined_scores[0]
#   # return the highest score element for each input in the batch
#   result = []
#   for datapoint in combined_scores:
#     combined_score = datapoint[0]
#     hyp, comb_score, w_score, lm_score = combined_score
#     result.append(hyp)
#   # result = [hyp for hyp, comb_score, w_score, lm_score in combined_scores]
#   #return text
#   return result

In [20]:
def decode_shallow_fusion_nbest(model, mel, best_of, lm_path, temperature, lm_weight, debug=False):
  """
    this function performs shallow fusion using best of N hypothesis (greedy decoding)
    by combining the scores of whisper and the language model score (which gets weighted by the lm_weight factor)

    - model: The Whisper model used for transcription.
    - mel: Represents Mel-spectrogram data, likely input features for the model.
    - best_of: An integer specifying the number of best hypotheses to consider during decoding.
    - lm_path: A string representing the path to the language model used for fusion.
    - temperature: A numerical value indicating the temperature parameter used during sampling. Higher temperature corresponds to more variation in the n best list
    - lm_weight: A numerical value representing the weight assigned to the language model scores during shallow fusion.
    - debug: A boolean flag (optional) indicating whether to print debug information (default is False).
      Useful when finding the optimal value for the lm_weight
  """
  # if testing with a single utterance without a dataloader
  if mel.shape[0] != 1:
    mel = torch.unsqueeze(mel, 0)
  options = whisper.DecodingOptions(fp16 = False, best_of=best_of, return_nbest=True, without_timestamps=True, temperature=temperature, language="hi")
  nbests = whisper.decode(model, mel, options)

  lm_model = kenlm.LanguageModel(lm_path)
  combined_scores = []

  for nbest in nbests:
        nbest_with_lm_score = [(c.text, c.avg_logprob, lm_model.score(c.text)) for c in nbest]
        normalized_scores = [(text,
                              whisper_score / len(text.split()),
                              lm_model.score(text) / len(text.split()))
                             for text, whisper_score, _ in nbest_with_lm_score]
        combined_score = [(text,
                           (whisper_score ** 2) + (lm_weight * lm_score ** 2),
                           whisper_score,
                           lm_score)
                          for text, whisper_score, lm_score in normalized_scores]
        combined_score.sort(key=lambda t: t[1], reverse=False)
        combined_scores.append(combined_score)
  if debug:
    print(combined_scores)
  # text, final_score, whisper_score, lm_score = combined_scores[0]
  # return the highest score element for each input in the batch
  result = []
  for datapoint in combined_scores:
    combined_score = datapoint[0]
    hyp, comb_score, w_score, lm_score = combined_score
    result.append(hyp)
  # result = [hyp for hyp, comb_score, w_score, lm_score in combined_scores]
  #return text
  return result

In [ ]:
# hypotheses = []
# references = []

# lm_path = '/content/language_model_3p0.bin'
# # # decoding parameters you can try playing around with to reach the optimal WER
# beam_size = 20
# lm_weight = 0.1
# whisper_results_with_shallow_fusion_beam_search = {}

# # audio = whisper.load_audio("/content/audio_1.wav")
# # audio = whisper.pad_or_trim(audio)
# # mel = whisper.log_mel_spectrogram(audio).to(model.device)
# # # decode_baseline decode_shallow_fusion_beam_search
# # result = decode_shallow_fusion_nbest(model, mel, best_of=best_of, lm_path=lm_path, temperature=temperature, lm_weight=lm_weight, debug=False)
# # result

# global audio_files
# # Transcribe each audio file
# for audio_file in audio_files:
#     audio = whisper.load_audio(audio_file)
#     audio = whisper.pad_or_trim(audio)
#     mel = whisper.log_mel_spectrogram(audio).to(model.device)
#     result = decode_shallow_fusion_beam_search(model, mel, beam_size=beam_size, lm_path=lm_path, lm_weight=lm_weight,debug=True)
#     whisper_results_with_shallow_fusion_beam_search[audio_file] = result

# # Print the results
# for audio_file, transcription in whisper_results_with_shallow_fusion_beam_search.items():
#     print(f"{audio_file}: {transcription}")

[[('पूजा में', -1.1214643695137718, -0.43847634575583716, -6.829880237579346), ('उगेउगे सूरज छठ पूजा में', -2.1856764639577557, -0.16355369937035344, -20.221227645874023), ('उगे सूरज छठ पूजा में', -2.1871442061204176, -0.2553261976975661, -19.318180084228516), ('उगेुगे सूरज छठ पूजा में', -2.2670618764815793, -0.24493911189417686, -20.221227645874023), ('उगे�गे सूरज छठ पूजा में', -2.277782917022705, -0.2556601524353027, -20.221227645874023), ('उगे उगे सूरज छठ पूजा में', -2.5255433553649533, -0.007628011126672068, -25.179153442382812), ('उगे होगे सूरज छठ पूजा में', -2.6245084639518494, -0.22054613790204447, -24.039623260498047), ('ओगे उगे सूरज छठ पूजा में', -2.7583050143334176, -0.20635970946281187, -25.519453048706055), ('ओगे ओगे सूरज छठ पूजा में', -2.7975419444422567, -0.21156667893932712, -25.859752655029297), ('ऊगे ऊगे सूरज छठ पूजा में', -2.8386132917096543, -0.14008843514227098, -26.985248565673828)]]
[[('हमारा नाम चंदन कुमार सिंघ्यार्म', -2.3871874144583036, -0.29216959982207324, -

In [26]:
hypotheses = []
references = []

best_of = 30
temperature = 0.9
lm_weight = .75

lm_path = '/content/language_model_3p0.bin'
# # decoding parameters you can try playing around with to reach the optimal WER


whisper_results_with_shallow_fusion_greedy_decoding = {}

# audio = whisper.load_audio("/content/audio_1.wav")
# audio = whisper.pad_or_trim(audio)
# mel = whisper.log_mel_spectrogram(audio).to(model.device)
# # decode_baseline decode_shallow_fusion_beam_search
# result = decode_shallow_fusion_nbest(model, mel, best_of=best_of, lm_path=lm_path, temperature=temperature, lm_weight=lm_weight, debug=False)
# result

global audio_files
# Transcribe each audio file
for audio_file in audio_files:
    audio = whisper.load_audio(audio_file)
    audio = whisper.pad_or_trim(audio)
    mel = whisper.log_mel_spectrogram(audio).to(model.device)
    result = decode_shallow_fusion_nbest(model, mel, best_of=best_of, lm_path=lm_path, temperature=temperature, lm_weight=lm_weight, debug=True)
    whisper_results_with_shallow_fusion_greedy_decoding[audio_file] = result

# Print the results
for audio_file, transcription in whisper_results_with_shallow_fusion_greedy_decoding.items():
    print(f"{audio_file}: {transcription}")

[[('उगे हुए सूरज छठ पूजा में', 8.297486779325695, -0.003981083951970583, -3.326153119405111), ('उगे हुए सूरज छठ पूजा में', 8.297486779325695, -0.003981083951970583, -3.326153119405111), ('उगे हुए सूरज छठ पूजा में', 8.297486779325695, -0.003981083951970583, -3.326153119405111), ('उगे हुए सूरज छठ पूजा में', 8.297486779330798, -0.003981084592880741, -3.326153119405111), ('उगे हुए सूरज छठ पूजा में', 8.297486779330798, -0.003981084592880741, -3.326153119405111), ('उगे हुए सूरज छठ पूजा में', 8.297486779392035, -0.0039810922838026475, -3.326153119405111), ('उगे हुए सूरज छठ पूजा में', 8.297486779392035, -0.0039810922838026475, -3.326153119405111), ('उगे उगे सूरज छठ पूजा में', 13.208156834468774, -0.006055265344599242, -4.196525573730469), ('उगे उगे सूरज छठ पूजा में', 13.208156834468774, -0.006055265344599242, -4.196525573730469), ('उगे उगे सूरज छठ पूजा में', 13.208156834530868, -0.006055270471880513, -4.196525573730469)]]
[[('हमारा नाम चंदन कुमार सिंह है और हम रांची से झारखण्ड स्टेडियम है', 6.

**Decoding with Default whisper **

In [ ]:
import whisper

# Load the model
model_orignal_whisper = whisper.load_model("small")

In [ ]:
##Code to perform call compute metrics on test dataset using our final model and calculate the WER and CER

hypotheses = []
references = []

lm_path = '/content/language_model_3p0.bin'

# decoding parameters you can try playing around with to reach the optimal WER
best_of = 10
temperature = 0.3
lm_weight = 0.01

for mel, text in tqdm(loader):
    # results = model.decode(mels, options)
    # hypotheses.extend([result.text for result in results])
    result = decode_shallow_fusion_nbest(model_orignal_whisper, mel, best_of=best_of, lm_path=lm_path, temperature=temperature, lm_weight=lm_weight, debug=False)
    hypotheses.extend(result)
    references.extend(text)


shallow_fusion_nbest_df = pd.DataFrame(dict(hypothesis=hypotheses, reference=references))
#shallow_fusion_nbest_df

100%|██████████| 11/11 [01:11<00:00,  6.46s/it]


In [ ]:
import string
import jiwer
from tqdm import tqdm

# Function to calculate WER and CER
def compute_metrics(model, loader, beam_size, lm_weight, lm_path):
    hypotheses = []
    references = []

    for mel, text in tqdm(loader):
        result = decode_shallow_fusion_beam_search(model, mel, beam_size=beam_size, lm_path=lm_path, lm_weight=lm_weight)
        hypotheses.extend(result)
        references.extend(text)

    extra_chars = '‘’“”।'
    reference = [text.translate(str.maketrans('', '', string.punctuation + extra_chars)) for text in references]
    hypo = [text.translate(str.maketrans('', '', string.punctuation + extra_chars)) for text in hypotheses]

    wer = jiwer.wer(reference, hypo)
    cer = jiwer.cer(reference, hypo)
    return wer, cer

# Tuning parameters
beam_sizes = [5, 10, 15, 20]
lm_weights = [0.01, 0.05, 0.1, 0.15,0.2,0.3,0.4,0.5]

best_wer = float('inf')
best_params = {'beam_size': None, 'lm_weight': None}

for beam_size in beam_sizes:
    for lm_weight in lm_weights:
        wer, cer = compute_metrics(model_orignal_whisper, loader, beam_size, lm_weight, lm_path)
        print(f"Beam size: {beam_size}, LM Weight: {lm_weight}, WER: {wer * 100:.2f} %, CER: {cer * 100:.2f} %")

        if wer < best_wer:
            best_wer = wer
            best_params['beam_size'] = beam_size
            best_params['lm_weight'] = lm_weight

# Print the best parameters
print(f"Best WER: {best_wer * 100:.2f} % with Beam Size: {best_params['beam_size']} and LM Weight: {best_params['lm_weight']}")


100%|██████████| 11/11 [00:37<00:00,  3.38s/it]


Beam size: 5, LM Weight: 0.01, WER: 87.06 %, CER: 60.38 %


100%|██████████| 11/11 [00:35<00:00,  3.26s/it]


Beam size: 5, LM Weight: 0.05, WER: 99.65 %, CER: 78.47 %


100%|██████████| 11/11 [00:34<00:00,  3.12s/it]


Beam size: 5, LM Weight: 0.1, WER: 90.91 %, CER: 65.59 %


100%|██████████| 11/11 [00:36<00:00,  3.30s/it]


Beam size: 5, LM Weight: 0.15, WER: 89.16 %, CER: 67.42 %


100%|██████████| 11/11 [00:34<00:00,  3.10s/it]


Beam size: 5, LM Weight: 0.2, WER: 89.86 %, CER: 63.41 %


100%|██████████| 11/11 [00:32<00:00,  2.99s/it]


Beam size: 5, LM Weight: 0.3, WER: 90.21 %, CER: 66.36 %


100%|██████████| 11/11 [00:32<00:00,  2.97s/it]


Beam size: 5, LM Weight: 0.4, WER: 90.91 %, CER: 65.66 %


100%|██████████| 11/11 [00:31<00:00,  2.84s/it]


Beam size: 5, LM Weight: 0.5, WER: 86.01 %, CER: 62.77 %


100%|██████████| 11/11 [00:55<00:00,  5.00s/it]


Beam size: 10, LM Weight: 0.01, WER: 82.52 %, CER: 58.62 %


100%|██████████| 11/11 [00:52<00:00,  4.82s/it]


Beam size: 10, LM Weight: 0.05, WER: 82.17 %, CER: 59.47 %


100%|██████████| 11/11 [00:53<00:00,  4.88s/it]


Beam size: 10, LM Weight: 0.1, WER: 84.27 %, CER: 58.90 %


100%|██████████| 11/11 [00:57<00:00,  5.20s/it]


Beam size: 10, LM Weight: 0.15, WER: 82.52 %, CER: 58.90 %


100%|██████████| 11/11 [00:58<00:00,  5.31s/it]


Beam size: 10, LM Weight: 0.2, WER: 84.97 %, CER: 60.94 %


100%|██████████| 11/11 [00:54<00:00,  5.00s/it]


Beam size: 10, LM Weight: 0.3, WER: 81.82 %, CER: 57.85 %


100%|██████████| 11/11 [00:59<00:00,  5.38s/it]


Beam size: 10, LM Weight: 0.4, WER: 81.12 %, CER: 55.17 %


100%|██████████| 11/11 [00:54<00:00,  4.97s/it]


Beam size: 10, LM Weight: 0.5, WER: 84.97 %, CER: 60.03 %


100%|██████████| 11/11 [01:21<00:00,  7.45s/it]


Beam size: 15, LM Weight: 0.01, WER: 84.97 %, CER: 63.27 %


100%|██████████| 11/11 [01:24<00:00,  7.72s/it]


Beam size: 15, LM Weight: 0.05, WER: 86.71 %, CER: 63.48 %


100%|██████████| 11/11 [01:19<00:00,  7.25s/it]


Beam size: 15, LM Weight: 0.1, WER: 85.66 %, CER: 64.18 %


100%|██████████| 11/11 [01:16<00:00,  6.94s/it]


Beam size: 15, LM Weight: 0.15, WER: 87.41 %, CER: 65.45 %


100%|██████████| 11/11 [01:10<00:00,  6.43s/it]


Beam size: 15, LM Weight: 0.2, WER: 86.36 %, CER: 67.21 %


100%|██████████| 11/11 [01:25<00:00,  7.77s/it]


Beam size: 15, LM Weight: 0.3, WER: 87.76 %, CER: 67.21 %


100%|██████████| 11/11 [01:13<00:00,  6.65s/it]


Beam size: 15, LM Weight: 0.4, WER: 86.01 %, CER: 66.71 %


100%|██████████| 11/11 [01:14<00:00,  6.81s/it]


Beam size: 15, LM Weight: 0.5, WER: 86.01 %, CER: 66.85 %


100%|██████████| 11/11 [01:41<00:00,  9.25s/it]


Beam size: 20, LM Weight: 0.01, WER: 88.46 %, CER: 72.48 %


100%|██████████| 11/11 [01:41<00:00,  9.19s/it]


Beam size: 20, LM Weight: 0.05, WER: 87.06 %, CER: 67.35 %


100%|██████████| 11/11 [01:41<00:00,  9.24s/it]


Beam size: 20, LM Weight: 0.1, WER: 89.16 %, CER: 72.41 %


100%|██████████| 11/11 [01:40<00:00,  9.11s/it]


Beam size: 20, LM Weight: 0.15, WER: 86.71 %, CER: 68.12 %


100%|██████████| 11/11 [01:41<00:00,  9.26s/it]


Beam size: 20, LM Weight: 0.2, WER: 86.36 %, CER: 68.33 %


100%|██████████| 11/11 [01:46<00:00,  9.72s/it]


Beam size: 20, LM Weight: 0.3, WER: 87.06 %, CER: 70.16 %


100%|██████████| 11/11 [01:43<00:00,  9.41s/it]


Beam size: 20, LM Weight: 0.4, WER: 87.76 %, CER: 73.33 %


100%|██████████| 11/11 [01:46<00:00,  9.67s/it]

Beam size: 20, LM Weight: 0.5, WER: 89.16 %, CER: 72.27 %
Best WER: 81.12 % with Beam Size: 10 and LM Weight: 0.4


WER: 159.79 %
CER: 100.28 %


# Decoding without LM as a baseline

In [21]:
##Code to perform call compute metrics on test dataset using our final model and calculate the WER and CER

hypotheses = []
references = []

# decoding parameters you can try playing around with to reach the optimal WER
beam_size = 5

for mel, text in tqdm(loader):
    # results = model.decode(mels, options)
    # hypotheses.extend([result.text for result in results])
    result = decode_baseline(model, mel, beam_size)
    hypotheses.extend(result)
    references.extend(text)

baseline_df = pd.DataFrame(dict(hypothesis=hypotheses, reference=references))
baseline_df

100%|██████████| 30/30 [00:34<00:00,  1.14s/it]


hypothesis  \
0                                  यही राम सब ही समझा   
1                                      राम शपथ में की   
2     यह गोरखपुर जनपद तथा गोरखपुर मंडल का मुख्यालय है   
3                          शायद दूसरे लोगों के काम आए   
4   विपीढ़िया सदस्य पूर्णिमा वर्मन और अरविंदन और व...   
5   पसंद आए तो आप उन्हें वापस पुराने रूप में ले सक...   
6                   हर को कम से कम एक श्रेणी में रखें   
7                                                  आप   
8   प्रधानमंत्री नरेन्द्र मोदी ने कहा है कि प्रधान...   
9   अन्य व्यंजन शब्द के अंत में आते हैं तो ज्यादात...   
10       आपकी सुविधा के लिए एक सूची मैं भी दे रहा हूँ   
11                  मैने पर्यटन में कुछ बदलाव किए हैं   
12                           भारत वचन सुनी देखी सुभाऊ   
13                   शाही परिवार के सदस्यों में से था   
14         यह जिला अभी भी वायु मार्ग से नहीं जुड़ा है   
15                           अफ्रीका के कुछ देशों में   
16                                   जान सुना सिख भाई   
17  वे लोग अपने मूल निवास स्थान की हिंदी की बोली ब...   
18                     इनके बारे में वह जानकारी भी है   
19                                               आपका   
20                                         ध्यान देना   
21  भारतवर्ष के उत्तरी राज्य उत्तर प्रदेश का एक नग...   
22            आज मैंने होली को निर्वाचित लेख बनाया है   
23              लगता है कि असल में दोनों भाग एक ही है   
24                     यह है इसलिए निम्न श्रेणी का है   
25                              वह शीर्षक शब्द ही रहा   
26                      ब्रह्मा विष्णु रुद्र सुर जेते   
27                          लेख जिन पर मैं काम करूंगा   
28                           भारत देश के उत्तर प्रदेश   
29            यद्यपि वहां से उन्होंने बीए पास कर लिया   

                                            reference  
0                           एहि बिधि राम सबहि समुझावा  
1                             राम सपथ मैं कीन्ह न काऊ  
2    यह गोरखपुर जनपद तथा गोरखपुर मण्डल का मुख्यालय है  
3                    शायद दूसरे नए लोगों के भी काम आए  
4   विकिपीडिया सदस्य पूर्णिमा वर्मन और अरविन्दन् औ...  
5   न पसंद आएँ तो आप उन्हें वापस पुराने रुप में ले...  
6                 हर लेख को कम से कम एक श्रेणी मे रखे  
7          आपकी पसंद के विषयों को जानकर प्रसन्नता हुई  
8             हम आपके सुन्दर भविष्य की कामना करते हैं  
9   अन्य व्यंजन जब शब्द के अंत में आते हैं तो ज़्य...  
10       आपकी सुविधा के लिये एक सूची मै भी दे रहा हूं  
11           मैंने पर्यटन भूगोल में कुछ बदलाव किए हैं  
12                            भरत बचन सुनि देखि सुभाऊ  
13                वह शाही परिवार के सदस्यों में से था  
14          यह जिला अभी भी वायु मार्ग से नहीं जुडा है  
15  अफ्रीका के कुछ देशों में सोलह मार्च को सूर्य क...  
16                         अस जियँ जानि सुनहु सिख भाई  
17  वे लोग अपने मूल निवास स्थान की हिन्दी की बोली ...  
18                   इनके बारे में वहां जानकारी भी है  
19           आप का योगदान वास्तव मै बहुत प्रशंसनीय है  
20  ध्यान दिलाने के लिए शुक्रिया मैने सुधार कर दिय...  
21  भारतवर्ष के उत्तरी राज्य उत्तर प्रदेश का एक नग...  
22             आज मैने होली को निर्वाचित लेख बनाया है  
23       लगता है कि असल में दोनों भाग एक ही अध्याय है  
24                यह हरा है इसलिये निम्न श्रेणी का है  
25                            वहाँ शीर्षक शब्द ही रहा  
26                      ब्रह्मा विष्णु रुद्र सुर जेते  
27                          लेख जिन पर मैं काम करूँगा  
28  भारत देश के उत्तर प्रदेश राज्य के आगरा जिले का...  
29          यद्यपि वहाँ से उन्होंने बी. ए पास कर लिया

In [ ]:

wer = jiwer.wer(list(baseline_df["reference"]), list(baseline_df["hypothesis"]))
cer = jiwer.cer(list(baseline_df["reference"]), list(baseline_df["hypothesis"]))

print(f"WER: {wer * 100:.2f} %")
print(f"CER: {cer * 100:.2f} %")

WER: 37.59 %
CER: 10.83 %


# Decoding with LM deep fusion

In [ ]:
# Download the binary LM from Gdrive
import gdown
url = "https://drive.google.com/uc?id=1-AspJVZRXcrFMuLKx8C4N7uo9PnQytcB"
output = "language_model_3p0.bin"
gdown.download(url, output, quiet=False)

Downloading...
From: https://drive.google.com/uc?id=1-AspJVZRXcrFMuLKx8C4N7uo9PnQytcB
To: /content/language_model_3p0.bin
100%|██████████| 280M/280M [00:03<00:00, 81.6MB/s]


'language_model_3p0.bin'

In [36]:
##Code to perform call compute metrics on test dataset using our final model and calculate the WER and CER

hypotheses = []
references = []

lm_path = '/content/language_model_3p0.bin'

# decoding parameters you can try playing around with to reach the optimal WER
beam_size = 20
lm_alpha = 0.5

for mel, text in tqdm(loader):
    # results = model.decode(mels, options)
    # hypotheses.extend([result.text for result in results])
    # result = decode_baseline(model, mel, beam_size)
    result = decode_deep_fusion(model, mel, beam_size=beam_size, lm_path=lm_path, lm_alpha=lm_alpha)
    hypotheses.extend(result)
    references.extend(text)

deep_fusion_df = pd.DataFrame(dict(hypothesis=hypotheses, reference=references))
deep_fusion_df

 20%|██        | 6/30 [04:47<19:11, 47.99s/it]


KeyboardInterrupt: ignored

# decoding parameters you can try playing around with to reach the optimal WER
beam_size = 15
lm_alpha = 0.5

WER: 37.24 %
CER: 10.83 %


# decoding parameters you can try playing around with to reach the optimal WER
beam_size = 15
lm_alpha = 1.5

WER: 37.24 %
CER: 10.83 %


In [ ]:
wer = jiwer.wer(list(deep_fusion_df["reference"]), list(deep_fusion_df["hypothesis"]))
cer = jiwer.cer(list(deep_fusion_df["reference"]), list(deep_fusion_df["hypothesis"]))

print(f"WER: {wer * 100:.2f} %")
print(f"CER: {cer * 100:.2f} %")

WER: 36.55 %
CER: 10.76 %


# Decoding with shallow fusion (Beam Search)

Downloading...
From: https://drive.google.com/uc?id=1-AspJVZRXcrFMuLKx8C4N7uo9PnQytcB
To: /content/language_model_3p0.bin
100%|██████████| 280M/280M [00:01<00:00, 199MB/s]


'language_model_3p0.bin'

In [39]:
##Code to perform call compute metrics on test dataset using our final model and calculate the WER and CER

hypotheses = []
references = []

lm_path = '/content/language_model_3p0.bin'

# # decoding parameters you can try playing around with to reach the optimal WER
beam_size = 10
lm_weight = 0.1

for mel, text in tqdm(loader):
    # results = model.decode(mels, options)
    # hypotheses.extend([result.text for result in results])
    result = decode_shallow_fusion_beam_search(model, mel, beam_size=beam_size, lm_path=lm_path, lm_weight=lm_weight)
    hypotheses.extend(result)
    references.extend(text)
    #exit()



100%|██████████| 30/30 [00:46<00:00,  1.54s/it]


In [40]:
import jiwer

shallow_fusion_beam_search_df = pd.DataFrame(dict(hypothesis=hypotheses, reference=references))
wer = jiwer.wer(list(shallow_fusion_beam_search_df["reference"]), list(shallow_fusion_beam_search_df["hypothesis"]))
cer = jiwer.cer(list(shallow_fusion_beam_search_df["reference"]), list(shallow_fusion_beam_search_df["hypothesis"]))

print(f"WER: {wer * 100:.2f} %")
print(f"CER: {cer * 100:.2f} %")

WER: 33.82 %
CER: 23.29 %


# Decoding with shallow fusion (Best of N hypothesis)

In [47]:
#after tuning

import pandas as pd
from tqdm import tqdm
import jiwer

# Function to calculate WER and CER
def compute_metrics(model, loader, best_of, temperature, lm_weight, lm_path):
    hypotheses = []
    references = []

    for mel, text in tqdm(loader):
        result = decode_shallow_fusion_nbest(model, mel, best_of=best_of, lm_path=lm_path, temperature=temperature, lm_weight=lm_weight, debug=False)
        hypotheses.extend(result)
        references.extend(text)

    shallow_fusion_nbest_df = pd.DataFrame(dict(hypothesis=hypotheses, reference=references))

    wer = jiwer.wer(shallow_fusion_nbest_df["reference"].tolist(), shallow_fusion_nbest_df["hypothesis"].tolist())
    cer = jiwer.cer(shallow_fusion_nbest_df["reference"].tolist(), shallow_fusion_nbest_df["hypothesis"].tolist())
    return wer, cer

# Tuning parameters
best_of_values = [5]
temperature_values = [0.1,0.25,0.5,0.75,0.9]
lm_weights = [0.1,0.25,0.5,0.75,0.9]

best_wer = float('inf')
best_params = {'best_of': None, 'temperature': None, 'lm_weight': None}

for best_of in best_of_values:
    for temperature in temperature_values:
        for lm_weight in lm_weights:
            wer, cer = compute_metrics(model, loader, best_of, temperature, lm_weight, lm_path)
            print(f"Best of: {best_of}, Temperature: {temperature}, LM Weight: {lm_weight}, WER: {wer * 100:.2f} %, CER: {cer * 100:.2f} %")

            if wer < best_wer:
                best_wer = wer
                best_params['best_of'] = best_of
                best_params['temperature'] = temperature
                best_params['lm_weight'] = lm_weight

# Print the best parameters
print(f"Best WER: {best_wer * 100:.2f} % with Best Of: {best_params['best_of']}, Temperature: {best_params['temperature']} and LM Weight: {best_params['lm_weight']}")


100%|██████████| 30/30 [00:30<00:00,  1.01s/it]


Best of: 5, Temperature: 0.1, LM Weight: 0.1, WER: 31.25 %, CER: 20.91 %


100%|██████████| 30/30 [00:30<00:00,  1.01s/it]


Best of: 5, Temperature: 0.1, LM Weight: 0.25, WER: 30.51 %, CER: 20.59 %


100%|██████████| 30/30 [00:30<00:00,  1.01s/it]


Best of: 5, Temperature: 0.1, LM Weight: 0.5, WER: 30.88 %, CER: 20.67 %


100%|██████████| 30/30 [00:30<00:00,  1.00s/it]


Best of: 5, Temperature: 0.1, LM Weight: 0.75, WER: 30.51 %, CER: 20.51 %


100%|██████████| 30/30 [00:30<00:00,  1.02s/it]


Best of: 5, Temperature: 0.1, LM Weight: 0.9, WER: 31.62 %, CER: 20.75 %


 73%|███████▎  | 22/30 [00:22<00:08,  1.04s/it]


KeyboardInterrupt: ignored

Best WER: 23.45 % with Best Of: 5, Temperature: 0.1 and LM Weight: 0.1

In [ ]:
# import pandas as pd
# from tqdm import tqdm
# import jiwer

# # Function to calculate WER and CER
# def compute_metrics(model, loader, best_of, temperature, lm_weight, lm_path):
#     hypotheses = []
#     references = []

#     for mel, text in tqdm(loader):
#         result = decode_shallow_fusion_nbest(model, mel, best_of=best_of, lm_path=lm_path, temperature=temperature, lm_weight=lm_weight, debug=False)
#         hypotheses.extend(result)
#         references.extend(text)

#     shallow_fusion_nbest_df = pd.DataFrame(dict(hypothesis=hypotheses, reference=references))

#     wer = jiwer.wer(shallow_fusion_nbest_df["reference"].tolist(), shallow_fusion_nbest_df["hypothesis"].tolist())
#     cer = jiwer.cer(shallow_fusion_nbest_df["reference"].tolist(), shallow_fusion_nbest_df["hypothesis"].tolist())
#     return wer, cer

# # Tuning parameters
# best_of_values = [5, 10, 15]
# temperature_values = [0.2, 0.3, 0.4]
# lm_weights = [0.01, 0.05, 0.1,0.2,0.3,0.4,0.5]

# best_wer = float('inf')
# best_params = {'best_of': None, 'temperature': None, 'lm_weight': None}

# for best_of in best_of_values:
#     for temperature in temperature_values:
#         for lm_weight in lm_weights:
#             wer, cer = compute_metrics(model, loader, best_of, temperature, lm_weight, lm_path)
#             print(f"Best of: {best_of}, Temperature: {temperature}, LM Weight: {lm_weight}, WER: {wer * 100:.2f} %, CER: {cer * 100:.2f} %")

#             if wer < best_wer:
#                 best_wer = wer
#                 best_params['best_of'] = best_of
#                 best_params['temperature'] = temperature
#                 best_params['lm_weight'] = lm_weight

# # Print the best parameters
# print(f"Best WER: {best_wer * 100:.2f} % with Best Of: {best_params['best_of']}, Temperature: {best_params['temperature']} and LM Weight: {best_params['lm_weight']}")


100%|██████████| 103/103 [04:30<00:00,  2.63s/it]


Best of: 5, Temperature: 0.2, LM Weight: 0.01, WER: 35.63 %, CER: 18.61 %


100%|██████████| 103/103 [04:31<00:00,  2.64s/it]


Best of: 5, Temperature: 0.2, LM Weight: 0.05, WER: 36.18 %, CER: 19.26 %


100%|██████████| 103/103 [04:29<00:00,  2.62s/it]


Best of: 5, Temperature: 0.2, LM Weight: 0.1, WER: 36.07 %, CER: 19.21 %


100%|██████████| 103/103 [04:29<00:00,  2.62s/it]


Best of: 5, Temperature: 0.2, LM Weight: 0.2, WER: 36.91 %, CER: 18.99 %


100%|██████████| 103/103 [04:29<00:00,  2.62s/it]


Best of: 5, Temperature: 0.2, LM Weight: 0.3, WER: 36.18 %, CER: 19.22 %


100%|██████████| 103/103 [04:29<00:00,  2.62s/it]


Best of: 5, Temperature: 0.2, LM Weight: 0.4, WER: 36.51 %, CER: 19.34 %


100%|██████████| 103/103 [04:29<00:00,  2.61s/it]


Best of: 5, Temperature: 0.2, LM Weight: 0.5, WER: 36.80 %, CER: 19.37 %


100%|██████████| 103/103 [04:30<00:00,  2.63s/it]


Best of: 5, Temperature: 0.3, LM Weight: 0.01, WER: 36.40 %, CER: 19.64 %


100%|██████████| 103/103 [04:31<00:00,  2.63s/it]


Best of: 5, Temperature: 0.3, LM Weight: 0.05, WER: 37.06 %, CER: 19.89 %


100%|██████████| 103/103 [04:29<00:00,  2.61s/it]


Best of: 5, Temperature: 0.3, LM Weight: 0.1, WER: 36.99 %, CER: 19.51 %


100%|██████████| 103/103 [04:31<00:00,  2.63s/it]


Best of: 5, Temperature: 0.3, LM Weight: 0.2, WER: 37.98 %, CER: 20.83 %


100%|██████████| 103/103 [04:29<00:00,  2.62s/it]


Best of: 5, Temperature: 0.3, LM Weight: 0.3, WER: 37.17 %, CER: 20.24 %


100%|██████████| 103/103 [04:30<00:00,  2.63s/it]


Best of: 5, Temperature: 0.3, LM Weight: 0.4, WER: 37.54 %, CER: 20.59 %


100%|██████████| 103/103 [04:29<00:00,  2.62s/it]


Best of: 5, Temperature: 0.3, LM Weight: 0.5, WER: 38.01 %, CER: 20.35 %


100%|██████████| 103/103 [04:31<00:00,  2.63s/it]


Best of: 5, Temperature: 0.4, LM Weight: 0.01, WER: 39.52 %, CER: 21.70 %


100%|██████████| 103/103 [04:30<00:00,  2.63s/it]


Best of: 5, Temperature: 0.4, LM Weight: 0.05, WER: 38.34 %, CER: 20.75 %


100%|██████████| 103/103 [04:30<00:00,  2.63s/it]


Best of: 5, Temperature: 0.4, LM Weight: 0.1, WER: 39.08 %, CER: 21.39 %


100%|██████████| 103/103 [04:31<00:00,  2.64s/it]


Best of: 5, Temperature: 0.4, LM Weight: 0.2, WER: 38.45 %, CER: 20.85 %


100%|██████████| 103/103 [04:32<00:00,  2.65s/it]


Best of: 5, Temperature: 0.4, LM Weight: 0.3, WER: 39.52 %, CER: 21.29 %


100%|██████████| 103/103 [04:32<00:00,  2.65s/it]


Best of: 5, Temperature: 0.4, LM Weight: 0.4, WER: 39.59 %, CER: 21.56 %


100%|██████████| 103/103 [04:33<00:00,  2.65s/it]


Best of: 5, Temperature: 0.4, LM Weight: 0.5, WER: 39.11 %, CER: 21.16 %


100%|██████████| 103/103 [06:53<00:00,  4.02s/it]


Best of: 10, Temperature: 0.2, LM Weight: 0.01, WER: 38.93 %, CER: 20.69 %


100%|██████████| 103/103 [06:52<00:00,  4.01s/it]


Best of: 10, Temperature: 0.2, LM Weight: 0.05, WER: 39.48 %, CER: 21.75 %


100%|██████████| 103/103 [06:54<00:00,  4.02s/it]


Best of: 10, Temperature: 0.2, LM Weight: 0.1, WER: 39.52 %, CER: 21.53 %


100%|██████████| 103/103 [06:54<00:00,  4.03s/it]


Best of: 10, Temperature: 0.2, LM Weight: 0.2, WER: 39.30 %, CER: 21.49 %


100%|██████████| 103/103 [06:54<00:00,  4.02s/it]


Best of: 10, Temperature: 0.2, LM Weight: 0.3, WER: 39.44 %, CER: 21.63 %


100%|██████████| 103/103 [06:52<00:00,  4.01s/it]


Best of: 10, Temperature: 0.2, LM Weight: 0.4, WER: 38.97 %, CER: 21.28 %


100%|██████████| 103/103 [06:54<00:00,  4.02s/it]


Best of: 10, Temperature: 0.2, LM Weight: 0.5, WER: 39.88 %, CER: 22.03 %


100%|██████████| 103/103 [06:55<00:00,  4.04s/it]


Best of: 10, Temperature: 0.3, LM Weight: 0.01, WER: 39.48 %, CER: 21.50 %


100%|██████████| 103/103 [06:59<00:00,  4.07s/it]


Best of: 10, Temperature: 0.3, LM Weight: 0.05, WER: 39.26 %, CER: 21.39 %


100%|██████████| 103/103 [06:54<00:00,  4.02s/it]


Best of: 10, Temperature: 0.3, LM Weight: 0.1, WER: 38.89 %, CER: 21.05 %


100%|██████████| 103/103 [06:55<00:00,  4.03s/it]


Best of: 10, Temperature: 0.3, LM Weight: 0.2, WER: 39.22 %, CER: 21.30 %


100%|██████████| 103/103 [06:56<00:00,  4.04s/it]


Best of: 10, Temperature: 0.3, LM Weight: 0.3, WER: 39.52 %, CER: 21.75 %


100%|██████████| 103/103 [06:55<00:00,  4.04s/it]


Best of: 10, Temperature: 0.3, LM Weight: 0.4, WER: 39.44 %, CER: 22.04 %


100%|██████████| 103/103 [06:54<00:00,  4.03s/it]


Best of: 10, Temperature: 0.3, LM Weight: 0.5, WER: 39.15 %, CER: 21.79 %


100%|██████████| 103/103 [06:56<00:00,  4.05s/it]


Best of: 10, Temperature: 0.4, LM Weight: 0.01, WER: 38.97 %, CER: 21.01 %


100%|██████████| 103/103 [06:57<00:00,  4.05s/it]


Best of: 10, Temperature: 0.4, LM Weight: 0.05, WER: 38.60 %, CER: 20.55 %


100%|██████████| 103/103 [07:00<00:00,  4.08s/it]


Best of: 10, Temperature: 0.4, LM Weight: 0.1, WER: 37.98 %, CER: 20.93 %


100%|██████████| 103/103 [06:55<00:00,  4.04s/it]


Best of: 10, Temperature: 0.4, LM Weight: 0.2, WER: 39.30 %, CER: 21.58 %


100%|██████████| 103/103 [06:57<00:00,  4.05s/it]


Best of: 10, Temperature: 0.4, LM Weight: 0.3, WER: 39.48 %, CER: 21.64 %


100%|██████████| 103/103 [06:56<00:00,  4.04s/it]


Best of: 10, Temperature: 0.4, LM Weight: 0.4, WER: 39.04 %, CER: 21.05 %


100%|██████████| 103/103 [07:01<00:00,  4.09s/it]


Best of: 10, Temperature: 0.4, LM Weight: 0.5, WER: 39.19 %, CER: 21.92 %


100%|██████████| 103/103 [09:24<00:00,  5.48s/it]


Best of: 15, Temperature: 0.2, LM Weight: 0.01, WER: 39.04 %, CER: 21.62 %


100%|██████████| 103/103 [09:27<00:00,  5.51s/it]


Best of: 15, Temperature: 0.2, LM Weight: 0.05, WER: 38.97 %, CER: 21.62 %


100%|██████████| 103/103 [09:25<00:00,  5.49s/it]


Best of: 15, Temperature: 0.2, LM Weight: 0.1, WER: 39.74 %, CER: 21.88 %


100%|██████████| 103/103 [09:28<00:00,  5.52s/it]


Best of: 15, Temperature: 0.2, LM Weight: 0.2, WER: 39.66 %, CER: 21.92 %


100%|██████████| 103/103 [09:25<00:00,  5.49s/it]


Best of: 15, Temperature: 0.2, LM Weight: 0.3, WER: 39.63 %, CER: 22.18 %


100%|██████████| 103/103 [09:27<00:00,  5.51s/it]


Best of: 15, Temperature: 0.2, LM Weight: 0.4, WER: 39.04 %, CER: 21.14 %


100%|██████████| 103/103 [09:25<00:00,  5.49s/it]


Best of: 15, Temperature: 0.2, LM Weight: 0.5, WER: 39.37 %, CER: 21.68 %


100%|██████████| 103/103 [09:30<00:00,  5.54s/it]


Best of: 15, Temperature: 0.3, LM Weight: 0.01, WER: 39.11 %, CER: 21.34 %


100%|██████████| 103/103 [09:31<00:00,  5.55s/it]


Best of: 15, Temperature: 0.3, LM Weight: 0.05, WER: 39.41 %, CER: 22.05 %


100%|██████████| 103/103 [09:29<00:00,  5.53s/it]


Best of: 15, Temperature: 0.3, LM Weight: 0.1, WER: 38.97 %, CER: 21.21 %


100%|██████████| 103/103 [09:29<00:00,  5.53s/it]


Best of: 15, Temperature: 0.3, LM Weight: 0.2, WER: 39.55 %, CER: 21.95 %


100%|██████████| 103/103 [09:36<00:00,  5.60s/it]


Best of: 15, Temperature: 0.3, LM Weight: 0.3, WER: 39.41 %, CER: 21.77 %


100%|██████████| 103/103 [09:29<00:00,  5.53s/it]


Best of: 15, Temperature: 0.3, LM Weight: 0.4, WER: 38.64 %, CER: 21.16 %


100%|██████████| 103/103 [09:30<00:00,  5.54s/it]


Best of: 15, Temperature: 0.3, LM Weight: 0.5, WER: 39.19 %, CER: 21.72 %


100%|██████████| 103/103 [09:32<00:00,  5.56s/it]


Best of: 15, Temperature: 0.4, LM Weight: 0.01, WER: 38.97 %, CER: 21.75 %


100%|██████████| 103/103 [09:33<00:00,  5.57s/it]


Best of: 15, Temperature: 0.4, LM Weight: 0.05, WER: 39.08 %, CER: 21.84 %


100%|██████████| 103/103 [09:38<00:00,  5.62s/it]


Best of: 15, Temperature: 0.4, LM Weight: 0.1, WER: 39.85 %, CER: 22.05 %


100%|██████████| 103/103 [09:32<00:00,  5.56s/it]


Best of: 15, Temperature: 0.4, LM Weight: 0.2, WER: 39.52 %, CER: 21.51 %


100%|██████████| 103/103 [09:31<00:00,  5.55s/it]


Best of: 15, Temperature: 0.4, LM Weight: 0.3, WER: 38.67 %, CER: 20.81 %


100%|██████████| 103/103 [09:36<00:00,  5.60s/it]


Best of: 15, Temperature: 0.4, LM Weight: 0.4, WER: 39.66 %, CER: 22.25 %


100%|██████████| 103/103 [09:32<00:00,  5.56s/it]

Best of: 15, Temperature: 0.4, LM Weight: 0.5, WER: 38.67 %, CER: 21.42 %
Best WER: 35.63 % with Best Of: 5, Temperature: 0.2 and LM Weight: 0.01


In [21]:
##Code to perform call compute metrics on test dataset using our final model and calculate the WER and CER

hypotheses = []
references = []

lm_path = '/content/language_model_3p0.bin'

# decoding parameters you can try playing around with to reach the optimal WER
#best_of = 10
#temperature = 0.3
#lm_weight = 0.01
#WER: 24.31 % , CER: 10.69 %

#best_of = 30 temperature = 0.5 lm_weight = .5 WER: 20.00 % CER: 6.89 %
#Best WER: 23.45 % with Best Of: 5, Temperature: 0.1 and LM Weight: 0.1
best_of = 5
temperature = 0.9
lm_weight = 0.75

#Best Of: 5, Temperature: 0.2 and LM Weight: 0.01

for mel, text in tqdm(loader):
    # results = model.decode(mels, options)
    # hypotheses.extend([result.text for result in results])
    hypotheses.extend(decode_shallow_fusion_nbest(model, mel, best_of=best_of, lm_path=lm_path, temperature=temperature, lm_weight=lm_weight, debug=False))
    references.extend(text)


shallow_fusion_nbest_df = pd.DataFrame(dict(hypothesis=hypotheses, reference=references))
#shallow_fusion_nbest_df




100%|██████████| 500/500 [09:11<00:00,  1.10s/it]


In [22]:
import string

wer = jiwer.wer(shallow_fusion_nbest_df["reference"].to_list(),shallow_fusion_nbest_df["hypothesis"].to_list())
cer = jiwer.cer(shallow_fusion_nbest_df["reference"].to_list(),shallow_fusion_nbest_df["hypothesis"].to_list())

print(f"WER: {wer * 100:.2f} %")
print(f"CER: {cer * 100:.2f} %")

WER: 13.62 %
CER: 8.86 %


graminVoice : WER: 30.07 %
CER: 14.57 %

IIITB
WER: 28.24 %
CER: 18.93 %

firefox : WER: 30.69 %
CER: 11.48 % with
best_of = 5
temperature = 0.9
lm_weight = 0.75

In [ ]:
import string

In [ ]:
extra_chars = '‘’“”।'
reference = [text.translate(str.maketrans('', '', string.punctuation + extra_chars)) for text in shallow_fusion_nbest_df["reference"].to_list()]
hypo = [text.translate(str.maketrans('', '', string.punctuation + extra_chars)) for text in shallow_fusion_nbest_df["hypothesis"].to_list()]

In [ ]:
reference[1]

'भारतवर्ष के उत्तरी राज्य उत्तर प्रदेश का एक नगर है'

In [ ]:
hypo[1]

'भारत वर्ष के उत्तरी राज्य उत्तर प्रदेश का एक नगर है'

In [ ]:
import string

wer = jiwer.wer(reference,hypo)
cer = jiwer.cer(reference,hypo)

print(f"WER: {wer * 100:.2f} %")
print(f"CER: {cer * 100:.2f} %")

WER: 28.24 %
CER: 18.72 %


In [ ]:
import string

wer = jiwer.wer(reference,hypo)
cer = jiwer.cer(reference,hypo)

print(f"WER: {wer * 100:.2f} %")
print(f"CER: {cer * 100:.2f} %")

WER: 26.67 %
CER: 19.05 %


In [ ]:
import string

wer = jiwer.wer(reference,hypo)
cer = jiwer.cer(reference,hypo)

print(f"WER: {wer * 100:.2f} %")
print(f"CER: {cer * 100:.2f} %")

WER: 37.41 %
CER: 18.30 %


# **Pointer 7 ## Final code t o push our model to hugging face **

In [ ]:
##Code to push the final model to hugging face

In [ ]:
kwargs = {
    "language": "hi",
    "model_name": "Whisper integrated LM 2011",  # a 'pretty' name for our model
    "finetuned_from": "openai/whisper-medium",
    "tasks": "automatic-speech-recognition",
    "tags": "hf-asr-leaderboard",
}

In [ ]:
model.push_to_hub(**kwargs)

AttributeError: ignored

**##Inferences of model vs BenchMarkModels -wave2wec and GoogleUSM**

**##Orignal  Whisper Model without any integration**

In [ ]:
!pip install openai-whisper


In [29]:
# List of audio files
audio_files = [
    "/content/audio_1_1.wav",
    "/content/audio_2.wav",
    "/content/audio_3_3.wav",
    "/content/audio_4_4.wav",
    "/content/audio_5_5.wav"
]

In [32]:
import whisper

# Load the model
model_orignal_whisper = whisper.load_model("small")  # Choose model size as needed



# Dictionary to hold results
orignal_whisper_results = {}

# Transcribe each audio file
for audio_file in audio_files:
    result = model_orignal_whisper.transcribe(audio_file, language="hi")
    orignal_whisper_results[audio_file] = result["text"]

# Print the results
for audio_file, transcription in orignal_whisper_results.items():
    print(f"{audio_file}: {transcription}")



/content/audio_1_1.wav:  उगे उगे सुद़ज चध पुजा में
/content/audio_2.wav:  अब राची से है, हमार आप चंदन कमार सिंग है, और जारक्ण्ट श्टेट है हमारा.
/content/audio_3_3.wav:  आँ आँ आँ सुबवाय में हम तुर लीट सुट हैं तर आप में लीट सुट हो एक हैं
/content/audio_4_4.wav:  बाहरत की वल्कुप में हाँसे सारे बहाग के बहुती दुखित है.
/content/audio_5_5.wav:  बिहार की श्थेती तोडी अच्छी नहीं हैं


[link text](https://)**##Orignal Integrated Whisper Model without fusion**

In [ ]:
#inferencing using audio files
##Code to perform call compute metrics on test dataset using our final model and calculate the WER and CER

hypotheses = []
references = []

lm_path = '/content/language_model_3p0.bin'

# # decoding parameters you can try playing around with to reach the optimal WER
beam_size = 20
lm_weight = 0.05
# Dictionary to hold results
orignal_whisper_results_without_fusion = {}

# audio = whisper.load_audio("/content/audio_1.wav")
# audio = whisper.pad_or_trim(audio)
# mel = whisper.log_mel_spectrogram(audio).to(model.device)
# options = whisper.DecodingOptions(fp16 = False, beam_size=5, without_timestamps=True, language="hi")
# result = whisper.decode(model, mel, options)
# baseline = result.text
# baseline

# Transcribe each audio file
for audio_file in audio_files:
    audio = whisper.load_audio(audio_file)
    audio = whisper.pad_or_trim(audio)
    mel = whisper.log_mel_spectrogram(audio).to(model.device)

    options = whisper.DecodingOptions(fp16=False, beam_size=5, without_timestamps=True, language="hi")
    result = whisper.decode(model, mel, options)
    baseline = result.text

    original_whisper_results_without_fusion[audio_file] = baseline

# Print the results
for audio_file, transcription in original_whisper_results_without_fusion.items():
    print(f"{audio_file}: {transcription}")


'हमारा नाम चंदन कुमार सिंह और हम राँची से झारखंड स्टेट याम'

**#Orignal Integrated Whisper Model with shallow fusion and beam search**

In [ ]:
hypotheses = []
references = []

lm_path = '/content/language_model_3p0.bin'
# # decoding parameters you can try playing around with to reach the optimal WER
beam_size = 20
lm_weight = 0.1
whisper_results_with_shallow_fusion_beam_search = {}

# audio = whisper.load_audio("/content/audio_1.wav")
# audio = whisper.pad_or_trim(audio)
# mel = whisper.log_mel_spectrogram(audio).to(model.device)
# # decode_baseline decode_shallow_fusion_beam_search
# result = decode_shallow_fusion_nbest(model, mel, best_of=best_of, lm_path=lm_path, temperature=temperature, lm_weight=lm_weight, debug=False)
# result

global audio_files
# Transcribe each audio file
for audio_file in audio_files:
    audio = whisper.load_audio(audio_file)
    audio = whisper.pad_or_trim(audio)
    mel = whisper.log_mel_spectrogram(audio).to(model.device)
    result = decode_shallow_fusion_beam_search(model, mel, beam_size=beam_size, lm_path=lm_path, lm_weight=lm_weight,debug=True)
    whisper_results_with_shallow_fusion_beam_search[audio_file] = result

# Print the results
for audio_file, transcription in whisper_results_with_shallow_fusion_beam_search.items():
    print(f"{audio_file}: {transcription}")

[-4.190035629272461, -3.0971128290349785, -3.387331008911133, -2.975449244181315, -3.098282814025879, -3.413821220397949, -3.4439942042032876, -3.385819435119629, -3.4123096466064453, -3.4189205169677734]


ValueError: ignored

**#Orignal Integrated Whisper Model with shallow fusion and greedy decoding:**

In [31]:
hypotheses = []
references = []

best_of = 5
temperature = 0.9
lm_weight = .75

lm_path = '/content/language_model_3p0.bin'
# # decoding parameters you can try playing around with to reach the optimal WER


whisper_results_with_shallow_fusion_greedy_decoding = {}

# audio = whisper.load_audio("/content/audio_1.wav")
# audio = whisper.pad_or_trim(audio)
# mel = whisper.log_mel_spectrogram(audio).to(model.device)
# # decode_baseline decode_shallow_fusion_beam_search
# result = decode_shallow_fusion_nbest(model, mel, best_of=best_of, lm_path=lm_path, temperature=temperature, lm_weight=lm_weight, debug=False)
# result

global audio_files
# Transcribe each audio file
for audio_file in audio_files:
    audio = whisper.load_audio(audio_file)
    audio = whisper.pad_or_trim(audio)
    mel = whisper.log_mel_spectrogram(audio).to(model.device)
    result = decode_shallow_fusion_nbest(model, mel, best_of=best_of, lm_path=lm_path, temperature=temperature, lm_weight=lm_weight, debug=True)
    whisper_results_with_shallow_fusion_greedy_decoding[audio_file] = result

# Print the results
for audio_file, transcription in whisper_results_with_shallow_fusion_greedy_decoding.items():
    print(f"{audio_file}: {transcription}")

[[('उगे हुए सूरज छठ पूजा में', 8.297486779677806, -0.003981128174771545, -3.326153119405111), ('उगे हुए सूरज छठ पूजा में', 8.29748677968291, -0.003981128815681704, -3.326153119405111), ('उगे हुए सूरज छठ पूजा में', 8.29748677973394, -0.003981135224783292, -3.326153119405111), ('उगे उगे सूरज छठ पूजा में', 13.208156833816789, -0.006055211508145896, -4.196525573730469), ('उगे उगे सूरज छठ पूजा में', 13.208156833863358, -0.006055215353606849, -4.196525573730469)]]
[[('अब रांची से है हमारा चंदन कुमार सिंह है झारखंड स्टेट है हमारा', 6.219314283628755, -0.002035517698482012, -2.8796551044170675), ('अब रांची से है हमारा चंदन कुमार सिंह है झारखण्ड स्टेट है हमारा', 7.361207824803155, -0.003030532541183325, -3.1328791104830227), ('अब रांची से है हमारा चंदन कुमार सिंह है झारखंड़ स्टेट है हमारा', 7.938000203439908, -0.0009298161892382146, -3.253305875338041), ('अब रांची से है हमारा चंदन कुमार सिंह है झारखंड़ स्टेट है हमारा', 7.938000203439908, -0.0009298161892382146, -3.253305875338041), ('अब रांची से 

**##Orignal Wave2wec**

In [ ]:
import soundfile as sf
from transformers import Wav2Vec2ForCTC, Wav2Vec2Tokenizer
import torch

# Load pre-trained model and tokenizer
tokenizer = Wav2Vec2Tokenizer.from_pretrained("facebook/wav2vec2-base-960h")
model = Wav2Vec2ForCTC.from_pretrained("facebook/wav2vec2-base-960h")

# Function to read and preprocess the audio file
def speech_file_to_array_fn(path):
    speech_array, sampling_rate = sf.read(path)
    return speech_array

# Function to perform inference
def asr_transcript(audio_file):
    # Load and preprocess the audio
    speech = speech_file_to_array_fn(audio_file)
    input_values = tokenizer(speech, return_tensors="pt", padding="longest").input_values

    # Perform inference
    with torch.no_grad():
        logits = model(input_values).logits

    # Get predicted IDs and decode them to text
    predicted_ids = torch.argmax(logits, dim=-1)
    transcription = tokenizer.batch_decode(predicted_ids)

    return transcription[0]

# Path to your audio file
audio_file = "/content/audio.wav"

# Perform inference and print the result
print("*******************wave 2 wec transcription***************** ")
print(asr_transcript(audio_file))


The tokenizer class you load from this checkpoint is not the same type as the class this function is called from. It may result in unexpected tokenization. 
The tokenizer class you load from this checkpoint is 'Wav2Vec2CTCTokenizer'. 
The class this function is called from is 'Wav2Vec2Tokenizer'.
/usr/local/lib/python3.10/dist-packages/transformers/models/wav2vec2/tokenization_wav2vec2.py:733: FutureWarning: The class `Wav2Vec2Tokenizer` is deprecated and will be removed in version 5 of Transformers. Please use `Wav2Vec2Processor` or `Wav2Vec2CTCTokenizer` instead.
  warnings.warn(
Some weights of the model checkpoint at facebook/wav2vec2-base-960h were not used when initializing Wav2Vec2ForCTC: ['wav2vec2.encoder.pos_conv_embed.conv.weight_g', 'wav2vec2.encoder.pos_conv_embed.conv.weight_v']
- This IS expected if you are initializing Wav2Vec2ForCTC from the checkpoint of a model trained on another task or with another architecture (e.g. initializing a BertForSequenceClassification mod

*******************wave 2 wec transcription***************** 
CHAMMUSHAA U CHACHA TRU I'L SHAR


**Orignal Google USM**

In [ ]:
#Google USM

!pip install google-cloud-speech


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 275.2/275.2 kB 4.4 MB/s eta 0:00:00


##Google USM Transcrption

In [ ]:
import os
os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = "/content/drivedownloaduplaod-eed71df0371c.json"


In [ ]:
!pip install librosa


In [ ]:
# import librosa
# import soundfile as sf

# # Load the audio file
# audio, sr = librosa.load('/content/audio_5.wav', sr=48000)  # Load with original sample rate (48kHz)

# # Resample to 16kHz
# audio_resampled = librosa.resample(audio, orig_sr=sr, target_sr=16000)

# # Save the resampled audio
# sf.write('/content/audio_5_5.wav', audio_resampled, 16000)


In [ ]:
input_file = '/content/audio_1.wav'
output_file = '/content/audio_1_1.wav'
original_sample_rate = 48000  # 48kHz
target_sample_rate = 16000    # 16kHz

In [ ]:
resample_audio(input_file, output_file, original_sample_rate, target_sample_rate)


In [ ]:
from pydub import AudioSegment

# Load the WAV file
wav_audio = AudioSegment.from_file("/content/audio_1_1.wav", format="wav")

# Export as FLAC
wav_audio.export("/content/audio_1_1.flac", format="flac")


<_io.BufferedRandom name='/content/audio_1_1.flac'>

In [ ]:
!curl -s -H "Content-Type: application/json" \
    -H "Authorization: Bearer "$(gcloud auth application-default print-access-token) \
    https://speech.googleapis.com/v1/speech:recognize \
    -d @sync-request.json

{
  "results": [
    {
      "alternatives": [
        {
          "transcript": "उदयवीर सूरज छठ पूजा में",
          "confidence": 0.80075234
        }
      ],
      "resultEndTime": "4.800s",
      "languageCode": "hi-in"
    }
  ],
  "totalBilledTime": "5s",
  "requestId": "2022486015774079337"
}


In [ ]:
!curl -s -H "Content-Type: application/json" \
    -H "Authorization: Bearer "$(gcloud auth application-default print-access-token) \
    https://speech.googleapis.com/v1/speech:recognize \
    -d @sync-request.json

{
  "results": [
    {
      "alternatives": [
        {
          "transcript": "मेरा नाम चंदन कुमार सिंह है और हम रांची से झारखंड स्टेट काम है",
          "confidence": 0.84342927
        }
      ],
      "resultEndTime": "6.750s",
      "languageCode": "hi-in"
    }
  ],
  "totalBilledTime": "7s",
  "requestId": "3225024303261540763"
}


In [ ]:
!curl -s -H "Content-Type: application/json" \
    -H "Authorization: Bearer "$(gcloud auth application-default print-access-token) \
    https://speech.googleapis.com/v1/speech:recognize \
    -d @sync-request.json

{
  "results": [
    {
      "alternatives": [
        {
          "transcript": "आज आज सुबह में हम फेवरेट स्वीट हैं कल रात में लेट सोए थे",
          "confidence": 0.80344194
        }
      ],
      "resultEndTime": "7.350s",
      "languageCode": "hi-in"
    }
  ],
  "totalBilledTime": "8s",
  "requestId": "2199201602030567921"
}


In [ ]:
!curl -s -H "Content-Type: application/json" \
    -H "Authorization: Bearer "$(gcloud auth application-default print-access-token) \
    https://speech.googleapis.com/v1/speech:recognize \
    -d @sync-request.json

{
  "results": [
    {
      "alternatives": [
        {
          "transcript": "भारत की वर्ल्ड कप में हाथ से सारे भाग के बहुत ही दुखित है",
          "confidence": 0.8286048
        }
      ],
      "resultEndTime": "6.830s",
      "languageCode": "hi-in"
    }
  ],
  "totalBilledTime": "7s",
  "requestId": "3756998355448913141"
}


In [ ]:
!curl -s -H "Content-Type: application/json" \
    -H "Authorization: Bearer "$(gcloud auth application-default print-access-token) \
    https://speech.googleapis.com/v1/speech:recognize \
    -d @sync-request.json

{
  "results": [
    {
      "alternatives": [
        {
          "transcript": "बिहार की स्थिति थोड़ी अच्छी नहीं है",
          "confidence": 0.65068185
        }
      ],
      "resultEndTime": "7.370s",
      "languageCode": "hi-in"
    }
  ],
  "totalBilledTime": "8s",
  "requestId": "5686012418812165510"
}


 ('हमारा नाम चंदन कुमार सिंह और हम राँजी से झारखंड स्टेट हैम',